### ISSUE TITLE
sendSMS examples

### ISSUE THREAD

**Issue by paulhayes** [neta540/ArduinoThreadedGSM#1]
Hi there, thanks for making this library. Could you add an sms send example. You mention that the string to pass to the method should be a PDU but it's not particularly clear what format this should take without an example.

**Comment by neta540** [comment]
The PDU format should be a hexadecimal string of the encoded PDU you would like to send.

This library uses SMS PDU mode. When sending a message, the message payload has to be encoded to PDU. In most cases that would require additional code or a library for encoding, and that does not belong here. Similarly, to parse or read an incoming message, you'd have to decode the received PDU. 

For information about PDU mode take a look at [http://www.gsm-modem.de/sms-pdu-mode.html](http://www.gsm-modem.de/sms-pdu-mode.html). This link has an example of sending a message in PDU mode and how to encode it. In their example **0011000B916407281553F80000AA0AE8329BFD4697D9EC37** will send a message "hellohello" to the phone number 0708251358.

**Comment by paulhayes** [comment]
Hi thanks for replying so quickly. Given the complexity and specificity of PDU encoding, would it not be prudent to incorporate a PDU parser and create an overloaded version of the sendSMS method which merely takes number and message as two char arrays? If I was to do this, would you be interested in merging it into your repo?

**Comment by neta540** [comment]
I chose not to integrate a PDU encoder/decoder module since I'd prefer this kind of module not to be specific only to this library. In my opinion, a universal library that is capable of generating a PDU message and parsing it, regardless of GSM library, device, or a modem model, would be much more beneficial to the Arduino community, than having it as a part of this specific library. It would be nice to see a good, resource-friendly PDU library written for Arduino devices. In case you were to write such library, more information about the sending process could be written in the readme, or even adding a new example, referencing to that library.

**Comment by paulhayes** [comment]
Thanks for your response Neta. I understand your decision. I'll be trying https://github.com/tardigrade888/c-pdu. If I have problems with that lib, I'll fork it and make it compatible with your methods.

I really like the way you've designed it. It's surprising to see so few people use it on platformio, I hope it gets more recognition.

**Comment by Piotr66** [comment]
Hi, I also thanks for making this library. Generally it’s working except sending SMS. I have read above about PDU and I have properly formatted message (I’ve check it on various online PDU decoder/encoder) but nothing happens. Maybe I’ve put it in wrong place ? Just below SIM900.setHandlers() in yours ArduinoThreadedGSM example. Could you add another example how to use sendSMS() function? Please… 
Ps. How to turn on debug mode in ThreadedGSM.h?

**Comment by neta540** [comment]
1.
Make sure your PDU message is correct, perhaps the receiver number is wrong. I can't remember how the receiver number has to be formatted. I suppose it has to be an international phone number, I can't remember at the moment whether a `+` sign prefix is necessary or should be omitted.

2.
Debug will be printed to serial when you add a line such as: 

```c++
#define THREADEDGSM_DEBUG Serial1
```

You will have to define it before including the library code with `#include <ThreadedGSM.h>`

As in the example: 

```c++
#define THREADEDGSM_DEBUG SerialDebug
#include <ThreadedGSM.h>
```

You will need to replace `SerialDebug` with the Serial to output debug information to.

**Comment by Piotr66** [comment]
Finally it’s working! There was two problems. 

The first one was a little funny. I’m using Visual Studio Code with PlatformIO extension and I install ThreadedGSM using it’s feature believing it’s the newest version. But as I discover later – it wasn’t.  So, believing it’s the newest version, I start to study Your code and I reach a conclusion that there is no way it can work (sending SMS).  First of all there was lack of “SMS.OutboxMsgContents = PDU;” statement in void sendSMS(String& PDU) function.  But sending SMS still wasn’t working. Then I “discover” on your Github page, that missing statement was already corrected by You, but some how I miss it. So, I download the “github version” of the ThreadedGSM, but it doesn’t help. 

The second problem -> In my case it was with the part of the Outbox() function which calculate TPDU_length (https://www.developershome.com/sms/cmgsCommand4.asp). In my case PDU always contain SMSC number in international format, so simply (SMS.OutboxMsgContents.length()-16) / 2 is doing the job.

Thank You for your effort and time, it spare a lot of my time. ;)

**Comment by neta540** [comment]
A fix seems to be needed here.

I don't like numbers like 16 appearing in the code when they can be of a different value than 16.

What was the value of `smsc_length` in your case? Can we use `smsc_length` to get the right formula for the final value?

You're welcome to create a PR with a fix which will work for all SMSC numbers and will calculate the 16 for you. maybe even `smsc_length` is 16 or 8, I am not sure. I don't have any board to test it, so please let me know.

Thank you.

**Comment by Piotr66** [comment]
> I don't like numbers like 16 appearing in the code

I don't like it too :) That's why I wrote "in my case".

> You're welcome to create a PR with a fix which will work for all SMSC

I don't have a needed knowledge to do this. But I think, that I figure it out and for You it'll be 1 minutes of coding.

SMS message encoded to PDU contain two parts - the SMSC part (contain SMSC number) and TPDU part (whatever it’s mean). We need to separate this two parts and count length of the TPDU part which is parameter to AT+CMGS command.

The SMSC part contain three sub-fields.

First octet is the length of the second and third sub-fields together.

Second sub-field it’s type of SMSC number and it may be 0x81 or 0x91. 
0x81 - the SMSC number assigned to the third sub-field can be either "85290000000"
            (country code included) or "90000000" (country code omitted)
0x91 - SMSC number assigned to the third sub-field should be "85290000000"
By the truth it doesn’t matter for us.

The third sub-field is SMSC number.

Here are two PDU’s with SMSC number and without:

With SMSC number
-------------------------
AT+CMGS=25
SMSC part		TPDU part
07 91 8498674523F1 11000B918421436587F90000AA0CC8329BFD06DDDF72363904

07 means that the second and third sub-fields together contain 14 hexadecimal digits in "918498674523F1" and each hexadecimal digit represents 4 bits. So, there are totally 7 octets. That's why the value of the first sub-field is 0x07. Plus length of “07” and we got 16 digits.

SMSC#: +48897654321
Receipient#:+48123456789
Validity:Rel 4d 
TP_PID:00
TP_DCS:00
TP_DCS-popis:Uncompressed Text
No class
Alphabet:Default

Hello world!
Length:12


Without SMSC number
-------------------------
AT+CMGS=25
SMSC part   TPDU part
00                 11000B918421436587F90000AA0CC8329BFD06DDDF72363904

There is nothing to explain I think.

SMSC#
Receipient:+48123456789
Validity:Rel 4d 
TP_PID:00
TP_DCS:00
TP_DCS-popis:Uncompressed Text
No class
Alphabet:Default

Hello world!
Length:12

**Comment by neta540** [comment]
Fixed in 0.9.2 

https://github.com/neta540/ArduinoThreadedGSM/commit/d1764c265faebbcb4f7d1daa37e04581818635ab

Please submit your results with 0.9.2.
Thank you.

**Comment by Piotr66** [comment]
With SMSC number given it's working OK, but without it - NO. 

With SMSC number given AT+CMGS=65 but without it - AT+CMGS=71 and this is wrong because TPDU part was the same in both cases.

**Comment by neta540** [comment]
I don't have a board right now to test it but I've ran the following code on two of your examples.

> 07918498674523F111000B918421436587F90000AA0CC8329BFD06DDDF72363904

And

> 0011000B918421436587F90000AA0CC8329BFD06DDDF72363904

```c++
#include <stdio.h>
#include <string.h>

int main(void) {
  const char pdu[] = "0011000B918421436587F90000AA0CC8329BFD06DDDF72363904";
  int smsc_length = 0;
  char smsc_len_b;

  for (int i = 0; i <= 1; i++) {
    smsc_len_b = pdu[i];
    smsc_length <<= 4;
    if (smsc_len_b >= '0' && smsc_len_b <= '9')
      smsc_length |= smsc_len_b - '0';
    else if (smsc_len_b >= 'a' && smsc_len_b <= 'f')
      smsc_length |= smsc_len_b - 'a' + 10;
    else if (smsc_len_b >= 'A' && smsc_len_b <= 'F')
      smsc_length |= smsc_len_b - 'A' + 10;
  }

  printf("pdu is: %s\n", pdu);
  printf("smsc_length: %d\n", smsc_length);
  int cmgs = (strlen(pdu) / 2 - (smsc_length + 1));
  printf("cmgs: %d\n", cmgs);

  return 0;
}
```

Both return:  **cmgs: 25**


I also ran several other tests for PDUs created using the encoder here: 
http://www.twit88.com/home/utility/sms-pdu-encode-decode
and calculations seems to be correct.

**Comment by Piotr66** [comment]
Forgive me please, I was wrong :(. I forgot that in function preparing PDU, I hard-coded 07 91 at the beginning, whether  SMSC number was given or not. I corrected this and it seems that everything is OK. SMS are sent in both cases. Sorry!

**Comment by neta540** [comment]
Great, then I close the issue.  Fixed in 0.9.2 https://github.com/neta540/ArduinoThreadedGSM/commit/d1764c265faebbcb4f7d1daa37e04581818635ab
